File: archived/v1/src/go/ch01/1.3/l1/modular_identities.go
package main

// normalize x into [0, m-1] for m > 0
func modNorm(x, m int64) int64 {
	r := x % m
	if r < 0 {
		r += m
	}
	return r
}

func ModAdd(a, b, m int64) int64 {
	a = modNorm(a, m)
	b = modNorm(b, m)
	return modNorm(a+b, m)
}

func ModSub(a, b, m int64) int64 {
	a = modNorm(a, m)
	b = modNorm(b, m)
	return modNorm(a-b, m)
}

func ModMul(a, b, m int64) int64 {
	a = modNorm(a, m)
	b = modNorm(b, m)
	// Use 128-bit math via big.Int if you expect overflow; for teaching:
	// do the reduction early when possible.
	return modNorm(a*b, m)
}

File: archived/v1/src/erlang/ch01/1.3/l1/src/modular_identities.erl
-module(modular_identities).
-export([mod_norm/2, mod_add/3, mod_sub/3, mod_mul/3]).

%% Normalize X into [0, M-1] for M > 0.
mod_norm(X, M) when M =< 0 ->
    erlang:error(badarg);
mod_norm(X, M) ->
    R = X rem M,
    case R < 0 of
        true  -> R + M;
        false -> R
    end.

mod_add(A, B, M) ->
    mod_norm(mod_norm(A, M) + mod_norm(B, M), M).

mod_sub(A, B, M) ->
    mod_norm(mod_norm(A, M) - mod_norm(B, M), M).

mod_mul(A, B, M) ->
    mod_norm(mod_norm(A, M) * mod_norm(B, M), M).